File: examples/starlette/tests/test_conformance.py
"""
Conformance tests for Starlette's multipart/form-data parsing.

This module runs all test cases from the parent tests/ directory against
a live Starlette/Uvicorn server to verify Starlette's multipart parsing behavior.
"""

import base64
import json
import socket
import subprocess
import sys
import time
from pathlib import Path

import httpx
import pytest

# Path to the conformance test suite
TESTS_ROOT = Path(__file__).parent.parent.parent.parent / "tests"


def get_unused_port():
    """Find an unused TCP port on localhost."""
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind(("127.0.0.1", 0))
        s.listen(1)
        port = s.getsockname()[1]
    return port


def collect_test_cases():
    """
    Collect all test cases from the tests/ directory.

    Returns a list of tuples (test_id, test_dir) for parametrization.
    """
    test_cases = []

    for category_dir in sorted(TESTS_ROOT.iterdir()):
        if not category_dir.is_dir():
            continue

        for test_dir in sorted(category_dir.iterdir()):
            if not test_dir.is_dir():
                continue

            # Verify this is a valid test case (has required files)
            test_json = test_dir / "test.json"
            headers_json = test_dir / "headers.json"
            input_raw = test_dir / "input.raw"

            if test_json.exists() and headers_json.exists() and input_raw.exists():
                test_cases.append((test_dir.name, test_dir))

    return test_cases


TEST_CASES = collect_test_cases()


@pytest.fixture(scope="session")
def server_url():
    """
    Start the Starlette/Uvicorn server on an unused port for the test session.

    Returns the base URL of the running server.
    """
    port = get_unused_port()

    # Path to the app
    app_path = Path(__file__).parent.parent / "src" / "app.py"

    # Start uvicorn server in a subprocess
    proc = subprocess.Popen(
        [
            sys.executable, "-m", "uvicorn",
            "src.app:app",
            "--host", "127.0.0.1",
            "--port", str(port),
        ],
        cwd=str(app_path.parent.parent),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    # Wait for server to start
    base_url = f"http://127.0.0.1:{port}"
    max_attempts = 50
    for attempt in range(max_attempts):
        try:
            response = httpx.get(f"{base_url}/health", timeout=1.0)
            if response.status_code == 200:
                break
        except (httpx.ConnectError, httpx.ReadTimeout):
            time.sleep(0.1)
    else:
        proc.terminate()
        stdout, stderr = proc.communicate()
        raise RuntimeError(
            f"Server failed to start.\nstdout: {stdout.decode()}\nstderr: {stderr.decode()}"
        )

    yield base_url

    # Cleanup: terminate the server
    proc.terminate()
    proc.wait()


def load_test_case(test_dir: Path):
    """Load test case data from a test directory."""
    test_json = json.loads((test_dir / "test.json").read_text())
    headers_json = json.loads((test_dir / "headers.json").read_text())
    input_raw = (test_dir / "input.raw").read_bytes()

    return test_json, headers_json, input_raw


def compare_part(actual: dict, expected: dict, test_id: str):
    """
    Compare an actual parsed part with the expected result.

    Returns a list of mismatches, or empty list if all matches.
    """
    mismatches = []

    # Compare name
    if actual.get("name") != expected.get("name"):
        mismatches.append(
            f"name mismatch: got {actual.get('name')!r}, expected {expected.get('name')!r}"
        )

    # Compare filename (null/None handling)
    actual_filename = actual.get("filename")
    expected_filename = expected.get("filename")
    if actual_filename != expected_filename:
        mismatches.append(
            f"filename mismatch: got {actual_filename!r}, expected {expected_filename!r}"
        )

    # Compare content_type (may be implementation-specific)
    actual_ct = actual.get("content_type")
    expected_ct = expected.get("content_type")
    # Only compare if expected specifies a content_type
    if expected_ct is not None and actual_ct != expected_ct:
        mismatches.append(
            f"content_type mismatch: got {actual_ct!r}, expected {expected_ct!r}"
        )

    # Compare body content
    if "body_text" in expected and expected["body_text"] is not None:
        actual_body = actual.get("body_text")
        expected_body = expected["body_text"]
        if actual_body != expected_body:
            mismatches.append(
                f"body_text mismatch: got {actual_body!r}, expected {expected_body!r}"
            )
    elif "body_base64" in expected and expected["body_base64"] is not None:
        # Binary content - compare as base64
        actual_b64 = actual.get("body_base64")
        actual_text = actual.get("body_text")
        expected_b64 = expected["body_base64"]

        # Server might return text if it happened to decode as UTF-8
        if actual_text is not None:
            actual_bytes = actual_text.encode("utf-8")
            actual_b64 = base64.b64encode(actual_bytes).decode("ascii")

        if actual_b64 != expected_b64:
            mismatches.append(
                f"body_base64 mismatch: got {actual_b64[:50]}..., expected {expected_b64[:50]}..."
            )

    # Compare body_size
    if "body_size" in expected:
        actual_size = actual.get("body_size")
        expected_size = expected["body_size"]
        if actual_size != expected_size:
            mismatches.append(
                f"body_size mismatch: got {actual_size}, expected {expected_size}"
            )

    return mismatches


@pytest.mark.parametrize("test_id,test_dir", TEST_CASES, ids=[tc[0] for tc in TEST_CASES])
def test_multipart_parsing(server_url: str, test_id: str, test_dir: Path):
    """
    Test Starlette's multipart parsing against a conformance test case.
    """
    test_json, headers_json, input_raw = load_test_case(test_dir)
    expected = test_json["expected"]

    # Build headers for the request
    headers = {}
    for key, value in headers_json.items():
        headers[key] = value

    # Send the raw multipart body to the server
    response = httpx.post(
        f"{server_url}/parse",
        content=input_raw,
        headers=headers,
        timeout=10.0,
    )

    result = response.json()

    # Check validity
    if expected.get("valid", True):
        # Test expects successful parsing
        assert result.get("valid") is True, (
            f"Expected valid parsing for {test_id}, but got: {result}"
        )

        expected_parts = expected.get("parts", [])
        actual_parts = result.get("parts", [])

        assert len(actual_parts) == len(expected_parts), (
            f"Part count mismatch for {test_id}: "
            f"got {len(actual_parts)}, expected {len(expected_parts)}"
        )

        all_mismatches = []
        for i, (actual_part, expected_part) in enumerate(zip(actual_parts, expected_parts)):
            mismatches = compare_part(actual_part, expected_part, test_id)
            if mismatches:
                all_mismatches.append(f"Part {i}: {'; '.join(mismatches)}")

        assert not all_mismatches, (
            f"Part comparison failed for {test_id}:\n" + "\n".join(all_mismatches)
        )
    else:
        # Test expects parsing to fail
        # Note: Starlette may still parse some "malformed" inputs successfully
        # due to lenient parsing, so we just check if the result makes sense
        if result.get("valid") is True:
            # Starlette successfully parsed what we expected to fail
            # This is OK - it's just more lenient than the spec requires
            pytest.skip(
                f"Starlette parsed malformed input {test_id} successfully "
                f"(lenient parsing)"
            )

File: examples/nodejs/tests/test_conformance.js
/**
 * Conformance tests for Node.js multipart/form-data parsing libraries.
 *
 * This module runs all test cases from the parent tests/ directory against
 * a live Express server to verify multipart parsing behavior for:
 * - multer
 * - formidable
 * - busboy
 */

import { describe, it, before, after } from 'node:test';
import assert from 'node:assert';
import { spawn } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import net from 'node:net';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Path to the conformance test suite
const TESTS_ROOT = path.resolve(__dirname, '..', '..', '..', 'tests');

// Parsers to test
const PARSERS = ['multer', 'formidable', 'busboy'];

/**
 * Find an unused TCP port.
 * @returns {Promise<number>}
 */
function getUnusedPort() {
  return new Promise((resolve, reject) => {
    const server = net.createServer();
    server.listen(0, '127.0.0.1', () => {
      const port = server.address().port;
      server.close(() => resolve(port));
    });
    server.on('error', reject);
  });
}

/**
 * Collect all test cases from the tests/ directory.
 * @returns {Array<{testId: string, testDir: string}>}
 */
function collectTestCases() {
  const testCases = [];

  const categories = fs.readdirSync(TESTS_ROOT);
  for (const category of categories.sort()) {
    const categoryPath = path.join(TESTS_ROOT, category);
    if (!fs.statSync(categoryPath).isDirectory()) continue;

    const tests = fs.readdirSync(categoryPath);
    for (const test of tests.sort()) {
      const testDir = path.join(categoryPath, test);
      if (!fs.statSync(testDir).isDirectory()) continue;

      // Verify this is a valid test case
      const testJson = path.join(testDir, 'test.json');
      const headersJson = path.join(testDir, 'headers.json');
      const inputRaw = path.join(testDir, 'input.raw');

      if (fs.existsSync(testJson) && fs.existsSync(headersJson) && fs.existsSync(inputRaw)) {
        testCases.push({ testId: test, testDir });
      }
    }
  }

  return testCases;
}

/**
 * Load test case data from a test directory.
 * @param {string} testDir
 * @returns {{testJson: Object, headersJson: Object, inputRaw: Buffer}}
 */
function loadTestCase(testDir) {
  const testJson = JSON.parse(fs.readFileSync(path.join(testDir, 'test.json'), 'utf-8'));
  const headersJson = JSON.parse(fs.readFileSync(path.join(testDir, 'headers.json'), 'utf-8'));
  const inputRaw = fs.readFileSync(path.join(testDir, 'input.raw'));

  return { testJson, headersJson, inputRaw };
}

/**
 * Compare an actual parsed part with the expected result.
 * @param {Object} actual
 * @param {Object} expected
 * @returns {string[]} List of mismatches
 */
function comparePart(actual, expected) {
  const mismatches = [];

  // Compare name
  if (actual.name !== expected.name) {
    mismatches.push(`name mismatch: got ${JSON.stringify(actual.name)}, expected ${JSON.stringify(expected.name)}`);
  }

  // Compare filename
  const actualFilename = actual.filename;
  const expectedFilename = expected.filename;
  if (actualFilename !== expectedFilename) {
    mismatches.push(`filename mismatch: got ${JSON.stringify(actualFilename)}, expected ${JSON.stringify(expectedFilename)}`);
  }

  // Compare content_type (only if expected specifies one)
  const actualCt = actual.content_type;
  const expectedCt = expected.content_type;
  if (expectedCt !== null && expectedCt !== undefined && actualCt !== expectedCt) {
    mismatches.push(`content_type mismatch: got ${JSON.stringify(actualCt)}, expected ${JSON.stringify(expectedCt)}`);
  }

  // Compare body content
  if (expected.body_text !== null && expected.body_text !== undefined) {
    if (actual.body_text !== expected.body_text) {
      mismatches.push(`body_text mismatch: got ${JSON.stringify(actual.body_text?.slice(0, 50))}, expected ${JSON.stringify(expected.body_text?.slice(0, 50))}`);
    }
  } else if (expected.body_base64 !== null && expected.body_base64 !== undefined) {
    // Binary content - compare as base64
    let actualB64 = actual.body_base64;
    if (actual.body_text !== null) {
      // Server might return text if it decoded as UTF-8
      actualB64 = Buffer.from(actual.body_text, 'utf-8').toString('base64');
    }
    if (actualB64 !== expected.body_base64) {
      mismatches.push(`body_base64 mismatch: got ${actualB64?.slice(0, 50)}..., expected ${expected.body_base64?.slice(0, 50)}...`);
    }
  }

  // Compare body_size
  if (expected.body_size !== undefined && actual.body_size !== expected.body_size) {
    mismatches.push(`body_size mismatch: got ${actual.body_size}, expected ${expected.body_size}`);
  }

  return mismatches;
}

/**
 * Check if a test is optional based on its tags.
 * @param {Object} testJson
 * @returns {boolean}
 */
function isOptionalTest(testJson) {
  const tags = testJson.tags || [];
  return tags.includes('optional');
}

/**
 * Start the server and return base URL.
 * @param {number} port
 * @returns {Promise<{proc: ChildProcess, baseUrl: string}>}
 */
async function startServer(port) {
  const appPath = path.resolve(__dirname, '..', 'src', 'app.js');
  const proc = spawn('node', [appPath], {
    cwd: path.dirname(appPath),
    env: { ...process.env, PORT: String(port) },
    stdio: ['ignore', 'pipe', 'pipe']
  });

  const baseUrl = `http://127.0.0.1:${port}`;

  // Wait for server to start
  const maxAttempts = 50;
  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      const response = await fetch(`${baseUrl}/health`);
      if (response.ok) {
        return { proc, baseUrl };
      }
    } catch {
      // Server not ready yet
    }
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  proc.kill();
  throw new Error('Server failed to start');
}

// Run tests
const TEST_CASES = collectTestCases();

describe('Multipart form-data conformance tests', async () => {
  let proc;
  let baseUrl;

  before(async () => {
    const port = await getUnusedPort();
    const result = await startServer(port);
    proc = result.proc;
    baseUrl = result.baseUrl;
  });

  after(() => {
    if (proc) {
      proc.kill();
    }
  });

  for (const parser of PARSERS) {
    describe(`${parser} parser`, () => {
      for (const { testId, testDir } of TEST_CASES) {
        it(`${testId}`, async (t) => {
          const { testJson, headersJson, inputRaw } = loadTestCase(testDir);
          const expected = testJson.expected;
          const optional = isOptionalTest(testJson);

          // Build headers for the request
          const headers = {};
          for (const [key, value] of Object.entries(headersJson)) {
            headers[key] = value;
          }

          // Send the raw multipart body to the server
          let result;
          try {
            const response = await fetch(`${baseUrl}/parse/${parser}`, {
              method: 'POST',
              body: inputRaw,
              headers
            });
            result = await response.json();
          } catch (error) {
            if (optional) {
              t.skip(`HTTP client rejected request for optional test: ${error.message}`);
              return;
            }
            throw new Error(`HTTP client error: ${error.message}`);
          }

          // Check validity
          if (expected.valid) {
            // Test expects successful parsing
            if (!result.valid) {
              if (optional) {
                t.skip(`Optional test failed: ${parser} could not parse input`);
                return;
              }
              throw new Error(`Expected valid parsing, but got error: ${result.error_message}`);
            }

            const expectedParts = expected.parts || [];
            const actualParts = result.parts || [];

            if (actualParts.length !== expectedParts.length) {
              if (optional) {
                t.skip(`Optional test: part count mismatch (got ${actualParts.length}, expected ${expectedParts.length})`);
                return;
              }
              throw new Error(`Part count mismatch: got ${actualParts.length}, expected ${expectedParts.length}`);
            }

            const allMismatches = [];
            for (let i = 0; i < actualParts.length; i++) {
              const mismatches = comparePart(actualParts[i], expectedParts[i]);
              if (mismatches.length > 0) {
                allMismatches.push(`Part ${i}: ${mismatches.join('; ')}`);
              }
            }

            if (allMismatches.length > 0) {
              if (optional) {
                t.skip(`Optional test failed:\n${allMismatches.join('\n')}`);
                return;
              }
              throw new Error(`Part comparison failed:\n${allMismatches.join('\n')}`);
            }
          } else {
            // Test expects parsing to fail
            if (result.valid) {
              // Parser successfully parsed what we expected to fail
              // This is OK - it's just more lenient than the spec requires
              t.skip(`${parser} parsed malformed input successfully (lenient parsing)`);
            }
          }
        });
      }
    });
  }
});
